### PR TITLE
[ORPO] Correct label mask for pad tokens

### PR DESCRIPTION
**Pull request by IlyaGusev**
Recent [fix](https://github.com/huggingface/trl/commit/57aebe9c36021e679662181468b04ff42432daf3) for calculating NLL loss for a whole sequence introduced a bug. When input_ids are copied to labels, pad tokens are not masked.

This PR aims to patch this by masking labels based on the attention mask.

I've tested it with batch_size=1, and now the initial loss is correct.

### CODE REVIEW

**Comment by kashif** [comment]
thanks @IlyaGusev can we replace all the magic `-100` by `self.label_pad_token_id` ?

File: trl/trainer/orpo_trainer.py [hunk 1]
```diff
@@ -714,6 +714,8 @@ def cross_entropy_loss(logits, labels):
             labels = concatenated_batch["concatenated_labels"].clone()
         else:
             labels = concatenated_batch["concatenated_input_ids"].clone()
+            attention_mask = concatenated_batch["concatenated_attention_mask"]
+            labels = torch.where(attention_mask == 1, labels, -100)
```
Comment (kashif) [inline trl/trainer/orpo_trainer.py:718]:
```suggestion
            labels = torch.where(attention_mask == 1, labels, self.label_pad_token_id)
```

**Comment by kashif** [comment]
cc @alvarobartt

File: trl/trainer/orpo_trainer.py (hunk 1 above)
Comment (IlyaGusev) [reply trl/trainer/orpo_trainer.py:718]: You are right, thank you!

**Review by kashif — APPROVED** [review]

**Review by younesbelkada — APPROVED** [review]

**Comment by AIR-hl** [comment]
> Recent [fix](https://github.com/huggingface/trl/commit/57aebe9c36021e679662181468b04ff42432daf3) for calculating NLL loss for a whole sequence introduced a bug. When input_ids are copied to labels, pad tokens are not masked.
> 
> This PR aims to patch this by masking labels based on the attention mask.
> 
> I've tested it with batch_size=1, and now the initial loss is correct.

@IlyaGusev Hi! With `0.8.6` version of the trainer, losses continue to decrease, but using current trainer losses fluctuate greatly, Is this normal?

**Comment by kashif** [comment]
@AIR-hl did you try to uncomment out the changes to confirm its because of this?

**Comment by AIR-hl** [comment]
> @AIR-hl did you try to uncomment out the changes to confirm its because of this?

@kashif I did some samll experiments, I tested `ORPO` with a sft `phi1_5` on different versions of `trl`（v0.8.6, v0.9.3）. For hyperparameters, I only changed the learning rate. I don't the model is bad, because for the `DPO` the reward accuracy can quickly stabilize at around 0.75

- `v0.8.6`; `lr=1e-5` 
351bc6ac9edb70606617f4d08adadc00

- `v0.9.3`; `lr=1e-5`
4e5e1efc162c1abc3ab7e8a425d89296

- `v0.9.3`; `lr=5e-5`
9ce6eccd35ca16790ea1b239cd0604c0

- `v0.9.3`; `lr=5e-7`
26db38217bf914e9c731e60d1471b024

my python script and launch file:
```
import multiprocessing
from dataclasses import dataclass, field
from typing import Optional, Dict

import torch
from datasets import load_dataset
from transformers import AutoModelForCausalLM, AutoTokenizer, TrainingArguments
from trl import (
    get_kbit_device_map,
    get_quantization_config,
    get_peft_config,
    ModelConfig,
    ORPOTrainer,
    ORPOConfig
)
from trl.commands.cli_utils import TrlParser


@dataclass
class ORPOScriptArguments:
    dataset_name: str = field(default=None, metadata={"help": "the dataset name"})
    dataset_train_split: str = field(default="train", metadata={"help": "The dataset split to use for training"})


if __name__ == "__main__":
    parser = TrlParser((ORPOScriptArguments, ORPOConfig, ModelConfig))
    args, training_args, model_config = parser.parse_args_and_config()

    torch_dtype = (
        model_config.torch_dtype
        if model_config.torch_dtype in ["auto", None]
        else getattr(torch, model_config.torch_dtype)
    )
    quantization_config = get_quantization_config(model_config)

    model_kwargs = dict(
        revision=model_config.model_revision,
        trust_remote_code=model_config.trust_remote_code,
        attn_implementation=model_config.attn_implementation,
        torch_dtype=torch_dtype,
        use_cache=False if training_args.gradient_checkpointing else True,
        device_map=get_kbit_device_map() if quantization_config is not None else None,
        quantization_config=quantization_config,
    )

    model = AutoModelForCausalLM.from_pretrained(model_config.model_name_or_path,
                                                 **model_kwargs)

    tokenizer = AutoTokenizer.from_pretrained(model_config.model_name_or_path)
    if tokenizer.pad_token is None:
        tokenizer.pad_token = tokenizer.eos_token
    tokenizer.padding_side = 'right'

    train_dataset = load_dataset(args.dataset_name,
                                 split=args.dataset_train_split, )
    train_dataset = train_dataset.shuffle(seed=training_args.seed)

    def process(row):
        row["prompt"] = tokenizer.apply_chat_template(row["prompt"], tokenize=False)
        row["chosen"] = tokenizer.apply_chat_template(row["chosen"], tokenize=False)
        row["rejected"] = tokenizer.apply_chat_template(row["rejected"], tokenize=False)
        return row

    train_dataset = train_dataset.map(
        process,
        num_proc=multiprocessing.cpu_count(),
        load_from_cache_file=False,
    )

    trainer = ORPOTrainer(
        model,
        args=training_args,
        train_dataset=train_dataset,
        tokenizer=tokenizer,
        peft_config=get_peft_config(model_config),
    )

    trainer.train()

    trainer.save_model(training_args.output_dir)
```

```
python orpo_version.py \
  --attn_implementation flash_attention_2 \
  --beta 0.1 \
  --bf16 \
  --dataset_name autodl-tmp/ultrafeedback-binarized-preferences-cleaned \
  --do_train \
  --gradient_checkpointing \
  --gradient_accumulation_steps 8 \
  --learning_rate 0.00005 \
  --lr_scheduler_type cosine \
  --logging_first_step \
  --logging_steps 5 \
  --max_length 2048 \
  --max_prompt_length 1024 \
  --max_steps -1 \
  --model_name_or_path autodl-tmp/phi-1_5-uc-01 \
  --num_train_epochs 1 \
  --output_dir autodl-tmp/output \
  --per_device_train_batch_size 4 \
  --report_to tensorboard \
  --save_strategy steps \
  --save_steps 200 \
  --save_total_limit 2 \
  --torch_dtype bfloat16 \
  --warmup_steps 20
```
